Limit CWND increase in slow start (#3232)
quicwg/base-drafts: draft-ietf-quic-recovery.md

congestion window, which is recommended to be the minimum of
10 * max_datagram_size and max(2* max_datagram_size, 14720)), where
max_datagram_size is the current maximum size of a datagram for the connection,
not including UDP or IP overhead. An implementation that does not use pacing

> SHOULD limit the congestion window increase during slow start to

Reviewer: In order to meet the previous criteria, how about "An implementation that doest not use pacing MAY limit the congestion window increase during slow start to 2 * max_datagram_size_ per ACK frame received {{?RFC3465}} in order to help limit bursts to the  initial congestion window."

Author: I think we can say either "MAY" or "SHOULD" instead of MAY/SHOULD. I am ok with changing it to "MAY" as there is already a requirement for limiting it to 10 datagrams in the previous line.

Reviewer 2: We've already said that a sender MUST limit bursts to IW, so there's no reason to repeat that. I would argue this is a SHOULD-level requirement; MAY is strange, since that's really a permission and that's not what we're going after here. I'd suggest dropping the phrase "in order to limit bursts" --- "An implementation that does not use pacing SHOULD limit the congestion window increase during slow start to 2 * max_datagram_size per ACK frame received {{?RFC3465}}. "

Author: sounds good to me. Updated it

congestion window, which is recommended to be the minimum of
10 * max_datagram_size and max(2* max_datagram_size, 14720)), where
max_datagram_size is the current maximum size of a datagram for the connection,

> not including UDP or IP overhead. Implementations not using a pacer SHOULD

Reviewer 2: ```suggestion
not including UDP or IP overhead. An implementation that does not use pacing SHOULD
```

       return
     if (congestion_window < ssthresh):
       // Slow start.
       if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in {{?RFC3465}}
         if (cwnd_increase < 2 * max_datagram_size):

>            cwnd_increase += bytes_acked

Reviewer: cwnd_increase doesn't appear to be used anywhere to change cwnd/etc.  Are more changes necessary?

Author: It is only used to limit the increase to 2*MSS in slow start

Reviewer 3: `bytes_acked` doesn't appear anywhere. I think it should be `acked_packet.size`?

Author: oops, that's correct. Fixing it

       return
     if (congestion_window < ssthresh):
       // Slow start.
       if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         if (cwnd_increase < 2 * max_datagram_size):
           cwnd_increase += bytes_acked

>            congestion_window += bytes_acked

Reviewer 3: Same issue as L1387 (`bytes_acked`)

Reviewer 3: I think this may allow increasing cwnd close to (3 x max_datagram_size).

e.g. max_datagram_size = 1000
ack 1 = 999 bytes
ack 2 = 999 bytes
ack 3 = 999 bytes

If we have 3 acks each acks 999 bytes, this code result in 999x3 bytes cwnd increase. 

To calculate remaining bytes correctly, I would propose the following:

```
  OnAckReceivedCC():
     cwnd_incr_max = 2 * max_datagram_size
...
  OnPacketAckedCC(acked_packet):
...
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         cwnd_incr = cwnd_incr_max - max(cwnd_incr_max - acked_packet.size, 0)
         congestion_window += cwnd_incr
         cwnd_incr_max -= cnwd_incr
```

Author: good point. How about the below:
```
 if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         if (cwnd_increase + acked_packet.size <=
             2 * max_datagram_size):
           cwnd_increase += acked_packet.size
           congestion_window += acked_packet.size
```

Author: There is a problem with my suggested code that the cwnd could end up being less than 2*MSS.

This is my updated proposal:
```
if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         if (cwnd_increase  < 2 * max_datagram_size):
           cwnd_increase = min(cwnd_increase + acked_packet.size,
                                            2 * max_datagram_size)
           congestion_window = cwnd_increase
```

Reviewer 3: > There is a problem with my suggested code that the cwnd could end up being less than 2*MSS.
> 
> This is my updated proposal:
> 
> ```
> if (pacing):
>          congestion_window += acked_packet.size
>        else:
>          // If pacing is not used, limit the increase to
>          // 2*max_datagram_size per ACK as described in
>          // {{?RFC3465}}
>          if (cwnd_increase  < 2 * max_datagram_size):
>            cwnd_increase = min(cwnd_increase + acked_packet.size,
>                                             2 * max_datagram_size)
>            congestion_window = cwnd_increase
> ```

I think it will reset congestion_window to a small value (<=2 mss).

Author: @junhochoi could you give an example where it would be less than 2MSS?

Reviewer 3: I mean this part:

> congestion_window = cwnd_increase

If you run my example above (3 acks 999 bytes each), cwin will be 2 x mss at the end with algorithms above.

I think you mean `+=`, not `=`, but if so it makes another issue (cwin will increase much faster than expected)

Reviewer 3: Last line is problematic because it will reset cwnd to cwnd_increase which is capped to 2 x mss.
```congestion_window = cwnd_increase```

Author: Isn't capping to 2*MSS the desired congestion window? `=` is my intended proposal to cap the congestion window to 2MSS.

Reviewer 3: I believe the basic intention of RFC3465 (section 2.3) is limiting cwnd *increase* during slow start per ACK,  (e.g. one ACK acknowledges 4 MSS --> only increase cwnd by 2 MSS, not 4 MSS), not resetting cwnd. If cwnd is reset to 2xMSS, it is basically same to idle timeout scenario.

Author: My bad. I would have to fix this.

Author: The below should work in all cases:
```
if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         if (cwnd_increase  < 2 * max_datagram_size):
           old_cwnd_increase = cwnd_increase
           cwnd_increase = min(cwnd_increase + acked_packet.size,
                                            2 * max_datagram_size)
           congestion_window += cwnd_increase - old_cwnd_increase
```

Reviewer 3: See mine at https://github.com/quicwg/base-drafts/pull/3232#discussion_r345506253
Basically same approach as yours. Note that I removed `if` and use `max()` instead of `min()`. I think either way works fine.

Author: I pushed my proposed changes before reading your comment. :-) I am fine with both approaches.
This issue will be discussed at IETF either today or tomorrow; we can decide it there.

       return
     if (congestion_window < ssthresh):
       // Slow start.
       if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to

>          // 2*max_datagram_size per ACK as described in

Reviewer 3: ```suggestion
         // 2 * max_datagram_size per ACK as described in
```
